File: day2/part1/main.py


LIMITS = {
        'RED': 12,
        'GREEN': 13,
        'BLUE': 14
}




def check_cubes(cubes):
    for cube in cubes.split(', '):
        [count, color] = cube.split(' ')
        if(int(count) > LIMITS[color.upper()]):
            return False
    return True

def check_game(game):
    for cubes in game.split('; '):
        if not check_cubes(cubes):
            return False
    return True


def process_line(line):
    line = line[:-1]
    split = line.split(': ')
    game_id = split[0].split(' ')[1]
    return game_id if check_game(split[1]) else 0

with open('input.txt', 'r') as f:
    count = 0;
    for line in f:
        count += int(process_line(line))

    print(count)

File: day2/part1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, err := os.Open("input.txt")

	if err != nil {
		fmt.Println(err)
		return
	}
	defer file.Close()

	var safe = 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		safe += ProcessLine(scanner.Text())
	}

	fmt.Println(safe)

}

func ProcessLine(line string) int {
	split := strings.Split(line, " ")
	var isIncreasing bool = false
	for i := 0; i < len(split)-1; i++ {
		i1, _ := strconv.Atoi(split[i])
		i2, _ := strconv.Atoi(split[i+1])

		if i == 0 {
			isIncreasing = i1 < i2
		}

		if isIncreasing && i2 <= i1 {
			return 0
		} else if !isIncreasing && i2 >= i1 {
			return 0
		}

		if Abs(i1-i2) > 3 {
			return 0
		}
	}
	return 1
}

func Abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
